### PR TITLE
Fix scale down logic

### PR DESCRIPTION
**Pull request by dmitry-ratushnyy**
The goal of this PR is to fix scale down logic of mongodb unit.

This is achieved by blocking unit removal until the unit is not removed from replica set configuration

### CODE REVIEW

**Review by delgod — APPROVED** [review]

File: src/charm.py [hunk 1]
```diff
@@ -214,16 +217,44 @@ def _on_start(self, event) -> None:
...
+        with MongoDBConnection(self.mongodb_config) as mongo:
+            try:
+                replset_members = mongo.get_replset_members()
+                return self.get_hostname_by_unit(self.unit.name) in replset_members
+            except NotReadyError as e:
+                logger.error(f"{self.unit.name}.is_unit_in_replica_set NotReadyError={e}")
+            except PyMongoError as e:
+                logger.error(f"{self.unit.name}.is_unit_in_replica_set PyMongoError={e}")
+
+    def _on_stop(self, event) -> None:
```
Comment (Mehdi-Bendriss) [inline src/charm.py:231]: I believe you could move this logic to the `storage_detaching` hook which is the first hook to be called on scaledown. And you already have some logic related to the scaledown there.

And do not use the `stop` hook (following conversation with @marcoppenheimer about upgrades, this hook gets called on a juju upgrade action - so better leave it for now)

File: src/charm.py [hunk 2]
```diff
@@ -214,16 +217,44 @@ def _on_start(self, event) -> None:
...
+        if "True" == self.unit_peer_data.get("unit_departed", "False"):
+            logger.debug(f"{self.unit.name} blocking on_stop")
+            is_in_replica_set = True
+            timeout = ON_STOP_TIMEOUT
```
Comment (Mehdi-Bendriss) [inline src/charm.py:235]: nit: you could use `tenacity.retry` to abstract the retry logic

File: tests/integration/test_teardown.py [hunk 3]
```diff
@@ -49,31 +49,9 @@ async def test_build_and_deploy(ops_test: OpsTest):
...
-        for count in scales:
-            await scale_and_verify(ops_test, count=count)
-
-
-@pytest.mark.skip("Issues with juju teardown")
-@pytest.mark.abort_on_fail
-async def test_scale_up_scale_down_large(ops_test: OpsTest):
-    """Scale up and down the application and verify the replica set is healthy."""
-    scales = [6, -6, 7, -7]
+    scales = [2, -1, -1, 2, -2, 3, -3, 4, -4, 5, -5, 6, -6, 7, -7]
```
Comment (Mehdi-Bendriss) [inline tests/integration/test_teardown.py:54]: Nice

File: src/charm.py (hunk 1 above)
Comment (dmitry-ratushnyy) [reply src/charm.py:231]: I am not sure about using `storage_detaching`, the usage of `stop` hook was suggested by juju devs. The idea is to block deletion of unit and the `stop` is the correct place to do it

File: src/charm.py (hunk 2 above)
Comment (dmitry-ratushnyy) [reply src/charm.py:235]: I don't think that using `renacity` here is the god idea

File: src/charm.py (hunk 1 above)
Comment (Mehdi-Bendriss) [reply src/charm.py:231]: `stop` comes later in the teardown process, `storage_detaching` is the entry point of teardown (you can confirm this by using `jhack utils tail -l 30 -d` then deleting the unit and watch the lifecycle) so it doesn't hurt to move the logic there imo. For reference: https://github.com/canonical/mongodb-operator/blob/main/src/charm.py#L155

Comment (dmitry-ratushnyy) [reply src/charm.py:231]: we still need to be sure what by the time app leader receives `RelationDepartedEvent` for a unit, the unit  and its pod is not deleted.

**Comment by dmitry-ratushnyy** [comment]
Closing as obsolete

**Pull request closed without merging**